src/engine.ts

```diff
@@ -210,6 +210,40 @@ export function initializePolyfill() {
     attributeOldValue: true,
   });
 
+  const originalAttachShadow = Element.prototype.attachShadow;
+  Element.prototype.attachShadow = function (options) {
+    const shadow = originalAttachShadow.apply(this, [options]);
+    mutationObserver.observe(shadow, {
```

Comment:
We probably want to move the `observe` to the `connect` handler of the controller, so we can `unobserve` during disconnect. Inside `attachShadow`, it should just be sufficient to call `getOrCreateInstance(shadow)`

Reply:
you mentioned `unobserve` in the `controller.disconnect`, should this be changed to the `ResizeObserver` or call `MutationObserver.disconnect` in the `controller.disconnect`?

Reply:
`MutationObserver` -- e.g. if the tree containing the `ShadowRoot` is disconnected, we want to clean it up as part of the disconnect.

In order to expose the `MutationObserver` to the `ShadowRootController`, I would recommend something like:

```ts
interface Disposable {
  (): void;
}

interface MutationObserverContext {
  observeMutation(node: Node): Disposable;
}
```

After the `MutationObserver` is created, just declare:

```
function observeMutation(node: Node) {
  mutationObserver.observe(node);
  return () => {
    mutationObserver.unobserve(node);
  };
}
```

Then pass this to the `ShadowRootController` similar to how other contexts are being passed. Then you can observe in `connected` and dispose the object in `disconnected` to clean up.

Reply:
Thank you for the context -- the implementation example is very helpful!

I am familiar with the pattern and why we need to implement it, but not as familiar with the nature of the `MutationObserver.disconnect` function. My curiosity was primarily around if `disconnect` would stop all mutation observations, or, if it would only stop the observation on the node the invocation is scoped in. (for example being able to target nodes with `ResizeObserver.unobserve(node)`
- I wasn't sure if there is a situation of only needing to stop observation on the `ShadowRoot` specifically.
- Or maybe in this particular implementation, it is ok to call `disconnect` during that life-cycle as we want to stop all observations and `observe` will be called again from the top down etc?

Reply:
D'oh. My bad. Of course `MutationObserver` doesn't have an `uobserve` function :) We'll want to create a new one for each `ShadowRoot` then (and reuse the same callback), where it'll be safe to call `disconnect()`.